### ISSUE TITLE
Print better error message when cuda driver library is not found by CMake

### ISSUE THREAD

**Issue by eyalroz** [arrayfire/arrayfire#1363]
After getting past other build issues with ArrayFire v3.3.1, I get:
```
CMake Error: The following variables are used in this project, but they are set to NOTFOUND.
Please set them or make sure they are set and tested correctly in the CMake files:
CUDA_CUDA_LIBRARY (ADVANCED)
    linked by target "afcuda" in directory /ufs/eyalroz/src/arrayfire-full-3.3.1/src/backend/cuda
    linked by target "write_cuda" in directory /ufs/eyalroz/src/arrayfire-full-3.3.1/test
    etc. etc.
```
If I add the following snippet to `CMakeLists.txt`:
```
    IF(NOT "$ENV{CUDA_CUDA_LIBRARY}" STREQUAL "")
      SET(CUDA_CUDA_LIBRARY $ENV{CUDA_CUDA_LIBRARY})
    ENDIF(NOT "$ENV{CUDA_CUDA_LIBRARY}" STREQUAL "")
```
and do `export CUDA_CUDA_LIBRARY=/path/to/cuda/lib64/libcudart.so`, then CMake happily (?) continues beyond this point. But that's not quite a solution...

PS - Sorry about filing lots of bugs on the same day... think about it as free QA on Fedora :-)

**Comment by pavanky** [comment]
@eyalroz Please try not to include stuff like `for no good reason`. Just because you do not understand what the error means does not mean there is no good reason for this failure. ArrayFire needs to link with `libcuda.so` which is obviously not present on your system. `libcudart.so` is `CUDA_CUDART_LIBRARY`. Pointing `CUDA_CUDA_LIBRARY` to it does not fix the issue.

`libcuda.so` is usually installed by the driver. Considering that you do not have a driver your machine, this error is expected. You could try copying `libcuda.so` from a machine that has the driver installed or try pointing it to `/path/to/cuda/lib64/stubs/libcuda.so`.

**Comment by eyalroz** [comment]
First - CUDA_CUDA_LIBRARY is an internal variable, not a missing environment variable. The user building ArrayFire should not need to know why that would not be set. So, I stand by my claim that this is "for no good reason". You could emit a message such as "Could not find libcuda.so" or "CUDA driver missing", and that would at least let us know what you were expecting.

**Comment by pavanky** [comment]
@eyalroz We are just using built in CMake Modules. We did not create this variable. Please forward your complaints in their direction.

**Comment by pavanky** [comment]
@eyalroz After thinking it over a little, it does seem like trying to print out a better error message might be useful in this particular case as it is not immediately obvious as other errors from CMake.

**Comment by eyalroz** [comment]
Maybe I should write them. But... I never had this problem. I mean, I write CUDA code (obviously not covering as much of the functionality as ArrayFire) and use CMake's FindCUDA module, and have not ever gotten this complaint just because the driver was unavailable.

Are you really sure the driver is necessary for _building_ ArrayFire, as opposed to _using_ it?

**Comment by pavanky** [comment]
@eyalroz one of the CUDA libraries we use brings in `libcuda.so` as a dependency. I do not remember which one.

**Comment by eyalroz** [comment]
When I copy libcuda.so manually from a machine with that file, I get:

```
[ 60%] Linking CXX executable image_processing/adaptive_thresholding_cuda
/usr/bin/ld: warning: libcuda.so.1, needed by ../src/backend/cuda/libafcuda.so.3.3.1, not found (try using -rpath or -rpath-link)
../src/backend/cuda/libafcuda.so.3.3.1: undefined reference to `cuLinkCreate_v2'
../src/backend/cuda/libafcuda.so.3.3.1: undefined reference to `cuLinkComplete'
../src/backend/cuda/libafcuda.so.3.3.1: undefined reference to `cuLaunchKernel'
../src/backend/cuda/libafcuda.so.3.3.1: undefined reference to `cuModuleLoadDataEx'
../src/backend/cuda/libafcuda.so.3.3.1: undefined reference to `cuModuleGetFunction'
../src/backend/cuda/libafcuda.so.3.3.1: undefined reference to `cuLinkAddData_v2'
```
So... do I need libcuda.so, libcuda.so.1, or both? If it's the latter, how do I tell CMake what to use when?

**Comment by shehzan10** [comment]
libcuda.so should simply be a simlink to libcuda.so.1. The runtime on all linux distribution loads the liblibrary.so.version file and not just the so file.

**Comment by shehzan10** [comment]
Fixed in hotfix-3.3.2.
